### PR TITLE
Add client metrics

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
Add client metrics implementation. Since the performance report generated with user activity, it should be better to reflect that behaviour and it would be much more realistic about the payload. I wanted to hear an early feedback whether this is a good approach.

I'll add `ClientFirstContentfulPaint`, `ClientLargestContentfulPaint` metrics in the meantime.

### CODE REVIEW

File: loadtest/control/simulcontroller/actions.go
Comment (agnivade) [reply loadtest/control/simulcontroller/actions.go:86]: Even better :stuck_out_tongue_closed_eyes:

File: loadtest/control/simulcontroller/controller.go
Comment (agarciamontoro) [inline loadtest/control/simulcontroller/controller.go:213]: Is this enough for sending it regularly? Again, I don't know how the browser does it, but if it follows a fixed frequency, maybe we should do something like that here?

File: loadtest/store/memstore/store.go
Comment (agarciamontoro) [inline loadtest/store/memstore/store.go:1215]: Do we need to do this inside the loop?

File: loadtest/store/memstore/store.go
Comment (agarciamontoro) [inline loadtest/store/memstore/store.go:1222]: Same here

File: loadtest/store/memstore/store.go
Comment (agarciamontoro) [inline loadtest/store/memstore/store.go:1234]: And same here

**Review by agarciamontoro — COMMENTED** [review]
Looks great, I like the approach! I left some comments, looking forward to seeing more metrics :eyes: Thanks for this work!

File: loadtest/control/simulcontroller/actions.go
Comment (isacikgoz) [reply loadtest/control/simulcontroller/actions.go:160]: TIL! We want to measure that whenever a user logs in (thinking of a scenario where opens a web page). But the user could also be logged in with the session so we may need to cover that scenario. The goal of the PR is to measure the load in the server with client metrics, but I think it could be a good idea to actually get realistic metrics from the agents. 

Maybe we can add this to backlog, WDYT?

File: loadtest/control/simulcontroller/controller.go
Comment (isacikgoz) [reply loadtest/control/simulcontroller/controller.go:213]: The browser should send this once in every minute but from my tests locally it's being sent whenever I focus to tab. So it doesn't send the metrics while I'm not using it.

File: loadtest/store/memstore/store.go
Comment (isacikgoz) [reply loadtest/store/memstore/store.go:1215]: Not necessarily, can take outside.

File: .golangci.yml
Comment (streamer45) [inline .golangci.yml:38]: Why is this commented out?

File: go.mod
Comment (streamer45) [inline go.mod:3]: Noting we are coupling this bump with the PR.

File: loadtest/control/simulcontroller/actions.go
Comment (streamer45) [inline loadtest/control/simulcontroller/actions.go:87]: Why are we dividing? The point of passing a float is that we can use seconds as unit as expected by most standard metrics.

**Review by streamer45 — COMMENTED** [review]
Thanks! Gave this a first quick pass and left some comments.

File: .golangci.yml
Comment (isacikgoz) [reply .golangci.yml:38]: Oh, that left from my attempt from testing revive and received plethora of errors. I can revert it, but `golint` also was not working anyway. 

```
WARN [linters_context] golint: This linter is fully inactivated: it will not produce any reports.
```

I can spend some time to set useful checks from revive and fix the complaints. Probably in a separate PR.

File: go.mod
Comment (isacikgoz) [reply go.mod:3]: Yeah, it's because the public package hence the server is bumped to `1.21`. Should we do it separately?

File: loadtest/control/simulcontroller/actions.go
Comment (isacikgoz) [reply loadtest/control/simulcontroller/actions.go:87]: Ops my bad, I directly copied the suggestion, it should've been `Milliseconds`, not `Seconds`.

File: loadtest/store/memstore/store.go
Comment (isacikgoz) [reply loadtest/store/memstore/store.go:1196]: This happens when you just copy instead of writing two lines 😓

File: go.mod
Comment (streamer45) [reply go.mod:3]: Up to you but it may be cleaner to split.

File: .golangci.yml
Comment (streamer45) [reply .golangci.yml:38]: That's fine but we should have a ticket if we defer it so we don't forget about it.

**Comment by agnivade** [comment]
@isacikgoz - Just checking on this, is there anything left other than us reviewing this again?

**Review by agnivade — APPROVED** [review]

**Review by streamer45 — APPROVED** [review]

File: loadtest/control/simulcontroller/actions.go
Comment (agarciamontoro) [inline loadtest/control/simulcontroller/actions.go:162]: I still think this is not semantically correct. I'm not sure if it's a problem with the naming (how are we using it exactly in the webapp?) or with the code here. For reference: I would understand this if the metric would be called something like `model.ClientTimeToLogin`.

Comment (agarciamontoro) [reply loadtest/control/simulcontroller/actions.go:162]: Also, answering to your comment in the other thread (sorry, just read it now):

>  TIL! We want to measure that whenever a user logs in (thinking of a scenario where opens a web page). But the user could also be logged in with the session so we may need to cover that scenario. The goal of the PR is to measure the load in the server with client metrics, but I think it could be a good idea to actually get realistic metrics from the agents.
Maybe we can add this to backlog, WDYT?

I'm ok adding a ticket to refine this to the backlog, but I'd still like to understand how the webapp uses this specific metric, I think I lack some context here, sorry.

Comment (isacikgoz) [reply loadtest/control/simulcontroller/actions.go:162]: An explanation can be found here: https://web.dev/articles/ttfb and turns out my interpretation is also incorrect. Just sent a canonical way of measuring it with your suggestion 😅

File: loadtest/user/userentity/report.go
Comment (isacikgoz) [reply loadtest/user/userentity/report.go:61]: Vanilla macos users are quite a lot among devs to be fair :p

File: loadtest/control/simulcontroller/actions.go
Comment (isacikgoz) [reply loadtest/control/simulcontroller/actions.go:162]: Oh right 🤦

File: loadtest/user/userentity/user.go
Comment (isacikgoz) [reply loadtest/user/userentity/user.go:115]: We are here to find that out :p, it shouldn't have much impact as it's just storing these observations locally and sending them in every minute.

**Review by agarciamontoro — APPROVED** [review]
Thank you, this is great! Do you think we can run a comparison before merge and check whether the overhead on the RoundTrip is negligible or not?

**Comment by isacikgoz** [comment]
> Thank you, this is great! Do you think we can run a comparison before merge and check whether the overhead on the RoundTrip is negligible or not?

That'd be my first loadtest tool comparison, is there a way to execute it?

**Comment by agnivade** [comment]
@isacikgoz - you can do a `ltctl comparison run`, or just run two consecutive bounded tests. Both works.

**Comment by agarciamontoro** [comment]
@isacikgoz: I think you'll have to run two consecutive bounded tests, updating the LoadTestDownloadURL between one and the other (you can use a local path pointing to the file generated by `make package` to use your changes). Let me know if you need some assistance and we can sync!

File: loadtest/user/userentity/report.go
Comment (agnivade) [inline loadtest/user/userentity/report.go:13]: Let's just use other/other for this. If we use all the platforms, then this becomes hard for a user to open a browser session and observe the "actual" client metrics in a live load test.

More context here: https://community.mattermost.com/core/pl/9y8cgyxxq7yq5yjorpqykbjkrw

Comment (M-ZubairAhmed) [reply loadtest/user/userentity/report.go:13]: Thank you for bringing this up. I think this approach will strike good balance between team wanting to use load testing tool for the web client metrics and tool being used to measure impact of web client metrics with load test.

**Comment by isacikgoz** [comment]
@agarciamontoro after discussing with @hmhealey I reverted the TTFB measurement to be made on login for only once. I think that more or less reflects our current usual scenario. I'll merge it as it is now if you are okay with it.

**Comment by agarciamontoro** [comment]
> @agarciamontoro after discussing with @hmhealey I reverted the TTFB measurement to be made on login for only once. I think that more or less reflects our current usual scenario. I'll merge it as it is now if you are okay with it.

Ok! I still don't think the name is correct, though, since it's not the time to first byte, but the time for the whole login to finish. Is there a way we can rename it? Don't get me wrong, I think that "the time for the whole login to finish" is a great metric to measure, it's just that "TTFB" means a very specific thing.

**Comment by isacikgoz** [comment]
@agarciamontoro Yes, you are right in terms of naming but this is just to simulate what does client metrics bring load. I'm not sure if we can use this metric for the agents anyway. We can add a new metric to simulate same payload but we would just add another time series in Prometheus. Is that something we should do?

**Comment by agarciamontoro** [comment]
Yeah, for this PR I'm ok with what we have as long as we don't deviate from the implementation in our clients. My concern is with the metric itself: if that's the name that we use in the clients, I think we should change it. But that's an off-topic for this PR, actually.

My TL;DR for this PR is: if the implementation here mimics what our clients do, then let's merge it (and keep it updated with the clients implementation in the future) :)

**Comment by isacikgoz** [comment]
@agarciamontoro gotcha, yes here this is only to mimic what our clients do. I'd say let's go with it for now and keep discussion on renaming it or using it per request etc.

**Comment by agarciamontoro** [comment]
Sounds good! All yours to merge, then :)

**Comment by hmhealey** [comment]
@isacikgoz @agarciamontoro I think there might be some misunderstanding here still. The TTFB measurement that the we app reports is indeed a proper TTFB, but it's the TTFB for the initial request made from the app to the server when loading the page, not the TTFB for every single request. It's sent exactly once on each page load.

There's more technical information on it and the other web app metrics here: https://mattermost.atlassian.net/wiki/spaces/ICU/pages/2715418659/Grafana+Metrics#Time-to-First-Byte

**Comment by agarciamontoro** [comment]
@hmhealey, can you point to the code measuring and reporting this?

**Comment by hmhealey** [comment]
@agarciamontoro The code that reports it is [here](https://github.com/mattermost/mattermost/blob/master/webapp/channels/src/utils/performance_telemetry/reporter.ts#L115), but the calculations are all done by either the browser itself or by [the Chrome team's `web-vitals` library](https://github.com/GoogleChrome/web-vitals). There's very little that we do here ourselves

**Comment by agarciamontoro** [comment]
Ah, ok ok, thanks, @hmhealey! Then I think we should move the login TTFB metric to the lower level possible, and get it for the actual first byte. If we measure it in `SimulController.login` as we're doing now, and we `defer` the measurement, we're adding several finished requests to that metric. `SimulController.login` calls `control.Login`, which calls `UserEntity.Login`, which in turn calls the final `Client4.Login`. That's where we need to measure it, I believe.
I agree that in terms of load it will be the same, but now that we're adding it, I would like to get it right. That metric may be useful in tests as well. @isacikgoz, thoughts?

**Comment by isacikgoz** [comment]
@agarciamontoro If you think that metric will be useful then let's go for it. So basically will run it again in the http trace but single time with sync.Once right?

**Comment by agnivade** [comment]
I think we are spending way too much time here. The purpose of this PR is to just add coverage for the client metrics feature. It should not be taken as the real data, because clients aren't going to be in the same network as the server. Where it will be useful is when someone actually logs in with a browser and points to the instance. That's why we are using "other"/"other" as the browser/platform combination for the load-test agent, so that we can clearly see real data when used from browsers.

It's already been months with this PR :)

**Comment by agarciamontoro** [comment]
I'm ok unblocking this PR and getting it merged. But I don't feel comfortable having a metric that doesn't do what its name suggests. So here's my proposal: let's merge this and, in parallel, create a ticket to address that concern, so that the tool comes closer to the real implementation. Thoughts?

**Comment by agnivade** [comment]
Sounds good to me. :+1:

**Comment by agarciamontoro** [comment]
@isacikgoz, feel free to merge whenever you want! We can create the ticket at any time :)

**Comment by isacikgoz** [comment]
@agarciamontoro I just didn't find the time to run another load test with these changes. Should we merge without it?

**Comment by agarciamontoro** [comment]
@isacikgoz A smoke test would be good just to ensure all is good. I can try to run one if you don't have the time

**Comment by isacikgoz** [comment]
@agarciamontoro It's not the time but generally I'm not so familiar with perf testing tool itself, if you have time I'd appreciate the help :D

**Comment by agarciamontoro** [comment]
Understandable! I'll run something quick as soon as I can, and maybe we can sync later at some point to do it together :)

**Comment by agarciamontoro** [comment]
@isacikgoz, can you fix the conflicts so I can test over a clean branch?

**Comment by isacikgoz** [comment]
@agarciamontoro It should be good now.

**Comment by agarciamontoro** [comment]
Just ran a dummy comparison and I can confirm it works perfectly: https://snapshots.raintank.io/dashboard/snapshot/ukK9uieg0hd1eca2S0caXwvZyCdwAKcM

As a bonus, here's a plot of the number of users against the number of reports submitted per minute, and it's the same, with exactly one minute of delay. Perfect alignment :)
image

**Pull request merged**
